### PR TITLE
Add bundle notification

### PR DESCRIPTION
**Pull request by LTA-Thinking**
## Description
Emits a notification when a bundle is done being processed

## Testing
New unit tests were added

## FHIR Team Checklist
- [x] **Update the title** of the PR to be succinct and less than 50 characters
- [x] **Add a milestone** to the PR for the sprint that it is merged (i.e. add S47)
- [x] Tag the PR with the type of update: **Bug**, **Dependencies**, **Enhancement**, or **New-Feature**
- [x] Tag the PR with Azure API for FHIR if this will release to the Azure API for FHIR managed service (CosmosDB or common code related to service)
- [x] Tag the PR with Azure Healthcare APIs if this will release to the Azure Healthcare APIs managed service (Sql server or common code related to service)
- [ ] CI is green before merge
- Review [squash-merge requirements](https://github.com/microsoft/fhir-server/blob/master/SquashMergeRequirements.md)

### Semver Change ([docs](https://github.com/microsoft/fhir-server/blob/master/docs/Versioning.md))
Patch|Skip|Feature|Breaking (reason)

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs [hunk 1]
```diff
@@ -217,7 +229,17 @@ public async Task<BundleResponse> Handle(BundleRequest request, CancellationToke
...
-                    return await ExecuteTransactionForAllRequests(responseBundle);
+                    var response = await ExecuteTransactionForAllRequests(responseBundle);
+
+                    int successfulRequestCount = 0;
+                    foreach (var entry in responseBundle.Entry)
+                    {
+                        successfulRequestCount += entry.Response.Status.StartsWith("2", StringComparison.InvariantCulture) ? 1 : 0;
+                    }
+
+                    await _mediator.Publish(new BundleMetricsNotification(successfulRequestCount, BundleType.Transaction), CancellationToken.None);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:240]: Do/should we care about the request types within the bundle? i.e.
{ "read", 2 },
{ "update", 5 }

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs [hunk 2]
```diff
@@ -217,7 +229,17 @@ public async Task<BundleResponse> Handle(BundleRequest request, CancellationToke
...
                     };
 
-                    return await ExecuteTransactionForAllRequests(responseBundle);
+                    var response = await ExecuteTransactionForAllRequests(responseBundle);
+
+                    int successfulRequestCount = 0;
+                    foreach (var entry in responseBundle.Entry)
+                    {
+                        successfulRequestCount += entry.Response.Status.StartsWith("2", StringComparison.InvariantCulture) ? 1 : 0;
+                    }
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:238]: Could use linq
```suggestion
                    successfulRequestCount = responseBundle.Entry.Count(x => x.Response.Status.StartsWith("2", StringComparison.Ordinal));
```

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs (hunk 1 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:240]: I don't believe we will be billing differently for different types of calls. @CaitlinV39, is this correct?

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs [hunk 3]
```diff
@@ -202,7 +204,13 @@ public async Task<BundleResponse> Handle(BundleRequest request, CancellationToke
                     };
 
                     await ExecuteAllRequests(responseBundle);
-                    return new BundleResponse(responseBundle.ToResourceElement());
+                    var response = new BundleResponse(responseBundle.ToResourceElement());
+
+                    int successfulRequestCount = responseBundle.Entry.Count(entry => entry.Response.Status.StartsWith("2", StringComparison.Ordinal));
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:209]: >entry.Response.Status.StartsWith("2" [](http://example.com/codeflow?start=85&length=36)

lol.

What do we return for delete request?

It's also so freaking brittle to do this... It's just a string with no restrictions, what if someone one day will come and change `200 Created` to `Created 200`?

Can we check for OperationOutcome with errors in response?

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleMetricsNotification.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+using static Hl7.Fhir.Model.Bundle;
+
+namespace Microsoft.Health.Fhir.Api.Features.Resources.Bundle
+{
+    public class BundleMetricsNotification : IMetricsNotification
+    {
+        public BundleMetricsNotification(int apiCalls, BundleType bundleType)
+        {
+            FhirOperation = bundleType == BundleType.Batch ? AuditEventSubType.Batch : AuditEventSubType.Transaction;
+            ResourceType = null;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleMetricsNotification.cs:17]: >null [](http://example.com/codeflow?start=27&length=4)

Why not  -Zoidberg-` Bundle`?

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs (hunk 1 above)
Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs:232]: To correct my previous comment, we should group by response code:

{ "200", 5 },
{ "429", 1 },
{ "500", 1 }

A mapping for these codes is done in the health platform.

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs (hunk 3 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleHandler.cs]: I changed the notification to send results sorted by response code. We will check for successful codes in PaaS.

**Review by Ivanidzo4ka — APPROVED** [review]
:shipit:

File: src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleMetricsNotification.cs [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+            FhirOperation = bundleType == BundleType.Batch ? AuditEventSubType.Batch : AuditEventSubType.Transaction;
+            ResourceType = KnownResourceTypes.Bundle;
+            ApiCallResults = apiCallResults;
+        }
+
+        public string FhirOperation { get; }
+
+        public string ResourceType { get; }
+
+        public IDictionary<string, int> ApiCallResults { get; }
```
Comment (Ivanidzo4ka) [inline src/Microsoft.Health.Fhir.Shared.Api/Features/Resources/Bundle/BundleMetricsNotification.cs:27]: >ApiCallResults [](http://example.com/codeflow?start=40&length=14)

I hope whoever takes this metric would knew how to parse it.

**Pull request merged**